### PR TITLE
feat: add `rewrite:RespHeaders` and modify the upstream response headers via `request` implementation

### PR DESCRIPTION
**Pull request by rampagecong**
### What this PR does / why we need it:
This PR add the ability to call upstream server when change the response header. 
When we implement SSO, we can set and modify cookie in APISIX plugin layer and also call upstream at the same time.

This is refer to https://github.com/apache/apisix-go-plugin-runner/issues/67

ext-plugin-proto : https://github.com/api7/ext-plugin-proto/pull/26
apisix-go-plugin-runner : https://github.com/apache/apisix-go-plugin-runner/pull/68

### Pre-submission checklist:

* [x] Did you explain what problem does this PR solve? Or what new features have been added?
* [x] Have you added corresponding test cases?
* [x] Have you modified the corresponding document?
* [x] Is this PR backward compatible? **If it is not backward compatible, please discuss on the [mailing list](https://github.com/apache/apisix/tree/master#community) first**

### CODE REVIEW

File: apisix/plugins/ext-plugin/init.lua [hunk 1]
```diff
@@ -611,6 +611,14 @@ local rpc_handlers = {
                 end
             end
 
+			local len = rewrite:RespHeadersLength()
```
Comment (tokers) [inline apisix/plugins/ext-plugin/init.lua:614]: Please reindent the codes.

File: apisix/plugins/ext-plugin/init.lua [hunk 2]
```diff
@@ -611,6 +611,14 @@ local rpc_handlers = {
                 end
             end
 
+            local len = rewrite:RespHeadersLength()
```
Comment (tokers) [inline apisix/plugins/ext-plugin/init.lua:614]: Should add some test cases.

Also, shall we exclude some vital headers such as `Connection`, `Content-Length`, `Transfer-Encoding` and etc.

**Comment by tzssangglass** [comment]
need to update

https://github.com/apache/apisix/blob/749094bb55421f5b715af18e376c373603bf5d62/rockspec/apisix-master-0.rockspec#L70

to 0.4.0

ref: https://github.com/api7/ext-plugin-proto/releases/tag/v0.4.0

File: apisix/plugins/ext-plugin/init.lua (hunk 2 above)
Comment (rampagecong) [reply apisix/plugins/ext-plugin/init.lua:614]: OK,I'll add some test cases.  
And exclude some important headers like Connection,Content-Length,Transfer-Encoding, Location,Server, WWW-Authenticate, Content-Encoding, Content-Type,Content-Location,Content-Type,Content-Language, feel free to give any other suggesions.

Comment (tokers) [reply apisix/plugins/ext-plugin/init.lua:614]: Let's see others' suggestions about this.

**Review by spacewander — COMMENTED** [review]
Let's add a test like:
https://github.com/apache/apisix/blob/80aaf7eefee6240df679a69422594f1bf2a70cac/t/plugin/ext-plugin/http-req-call.t#L210

You can prepare data for the mock runner in:
https://github.com/apache/apisix/blob/80aaf7eefee6240df679a69422594f1bf2a70cac/t/lib/ext-plugin.lua#L323

You can refer to https://github.com/apache/apisix/blob/master/docs/en/latest/internal/testing-framework.md for how to write/run test.

For example,
```
Check response headers.

--- response_headers
X-Resp: foo
X-Req: foo, bar
```

File: apisix/plugins/ext-plugin/init.lua (hunk 2 above)
Comment (spacewander) [reply apisix/plugins/ext-plugin/init.lua:614]: > Connection,Content-Length,Transfer-Encoding, Location,Server, WWW-Authenticate, Content-Encoding, Content-Type,Content-Location,Content-Type,Content-Language

LGTM

**Comment by rampagecong** [comment]
The test cases can't be run in both my two centos machines(follow the step in docs). So I have to try follow the exsisting cases and push code to ci to verify it.

`[root@localdev apisix]# make test
make test
[ info ] test -> [ Start ]
git submodule update --init --recursive
prove -I../test-nginx/lib -I./ -r -s t/
t/node/remote-addr.t ............................ nginx: [error] init_by_lua error: /data/service/test/apisix/deps/share/lua/5.1/protoc.lua:1148: attempt to call field 'load' (a nil value)
stack traceback:
	/data/service/test/apisix/deps/share/lua/5.1/protoc.lua:1148: in function </data/service/test/apisix/deps/share/lua/5.1/protoc.lua:1147>
	/data/service/test/apisix/deps/share/lua/5.1/protoc.lua:1192: in main chunk
	[C]: in function 'require'
	...vice/test/apisix/apisix/plugins/grpc-transcode/proto.lua:20: in main chunk
	[C]: in function 'require'
	/data/service/test/apisix/apisix/admin/proto.lua:23: in main chunk
	[C]: in function 'require'
	/data/service/test/apisix/apisix/admin/init.lua:51: in main chunk
	[C]: in function 'require'
	/data/service/test/apisix/apisix/init.lua:34: in main chunk
	[C]: in function 'require'
	init_by_lua:9: in main chunk
`

File: apisix/plugins/ext-plugin/init.lua [hunk 3]
```diff
@@ -612,10 +612,27 @@ local rpc_handlers = {
             end
 
             local len = rewrite:RespHeadersLength()
+            local excludeRespHeader = {
```
Comment (tokers) [inline apisix/plugins/ext-plugin/init.lua:615]: We can just use a hash-like to save these headers so we don't need to iterate all of them, i.e. we can reduce the time complicity from O(N^2) => O(N).

File: apisix/plugins/ext-plugin/init.lua [hunk 4]
```diff
@@ -611,6 +611,29 @@ local rpc_handlers = {
                 end
             end
 
+            local len = rewrite:RespHeadersLength()
+            local excludeRespHeader = {
```
Comment (tzssangglass) [inline apisix/plugins/ext-plugin/init.lua:615]:
```suggestion
            local exclude_resp_header = {
```

is better

File: t/lib/ext-plugin.lua [hunk 5]
```diff
@@ -67,7 +67,7 @@ function _M.go(case)
             ty = constants.RPC_ERROR
             err_resp.Start(builder)
             err_resp.AddCode(builder, err_code.BAD_REQUEST)
-            local req = prepare_conf_req.End(builder)
+                local req = prepare_conf_req.End(builder)
```
Comment (shuaijinchao) [inline t/lib/ext-plugin.lua:70]:
```suggestion
            local req = prepare_conf_req.End(builder)
```

File: t/lib/ext-plugin.lua [hunk 6]
```diff
@@ -367,6 +367,33 @@ function _M.go(case)
             local action = http_req_call_rewrite.End(builder)
             build_action(action, http_req_call_action.Rewrite)
 
+        elseif case.rewrite_resp_header == true then
+            local hdrs = {
```
Comment (tokers) [inline t/lib/ext-plugin.lua:371]: I think you can also add some unexpected headers and verify if they exist in the response headers.

**Comment by spacewander** [comment]
> `[root@iZbp1efn5vn52xspp2bid5Z apisix]# make test [ info ] test -> [ Start ] git submodule update --init --recursive prove -I../test-nginx/lib -I./ -r -s t/ t/stream-node/sni.t ............................. nginx: [error] init_by_lua error: error loading module 'ssl.core' from file '/opt/service/test/apisix/deps/lib/lua/5.1/ssl.so': /opt/service/test/apisix/deps/lib/lua/5.1/ssl.so: undefined symbol: sk_free stack traceback: [C]: at 0x7f6b5dc70a20 [C]: in function 'require' /opt/service/test/apisix/deps/share/lua/5.1/ssl.lua:8: in main chunk [C]: in function 'require' /opt/service/test/apisix/apisix/patch.lua:22: in main chunk [C]: in function 'require' /opt/service/test/apisix/apisix/init.lua:28: in main chunk [C]: in function 'require' init_by_lua:9: in main chunk`

@shuaijinchao 
Would you provide help to solve the OpenSSL problem?

**Comment by rampagecong** [comment]
`[root@iZbp1efn5vn52xspp2bid5Z apisix]# prove -I../test-nginx/lib -I./ -r -s t/plugin/ext-plugin/http-req-call.t
t/plugin/ext-plugin/http-req-call.t .. nginx: [warn] stream [lua] health_check.lua:99: report_failure(): update endpoint: http://127.0.0.1:2379 to unhealthy
nginx: [warn] stream [lua] v3.lua:151: _request_uri(): http://127.0.0.1:2379: connection refused. Retrying
nginx: [error] stream [lua] init.lua:822: stream_init(): failed to load the configuration: has no healthy etcd endpoint available
nginx: [warn] [lua] health_check.lua:99: report_failure(): update endpoint: http://127.0.0.1:2379 to unhealthy
nginx: [warn] [lua] v3.lua:151: _request_uri(): http://127.0.0.1:2379: connection refused. Retrying
nginx: [error] [lua] init.lua:86: http_init(): failed to load the configuration: has no healthy etcd endpoint available
t/plugin/ext-plugin/http-req-call.t .. 1/? 
#   Failed test 'TEST 7: rewrite args - status code ok'
#   at /opt/service/test/apisix/../test-nginx/lib/Test/Nginx/Socket.pm line 948.
#          got: '404'
#     expected: '200'

#   Failed test 'TEST 7: rewrite args - response_body - response is expected (repeated req 0, req 0)'
#   at /opt/service/test/apisix/../test-nginx/lib/Test/Nginx/Socket.pm line 1589.
# @@ -1,3 +1 @@
# -uri: /plugin_proxy_rewrite_args
# +{"error_msg":"404 Route Not Found"}
# -a: foo,bar
# -c: bar`

'TEST 7' also failed. Is there a problem with setting it?Please help me.
@shuaijinchao

**Comment by shuaijinchao** [comment]
@rampagecong r Ok, I need some time to check this question and get back to you later.

File: t/lib/ext-plugin.lua (hunk 6 above)
Comment (rampagecong) [reply t/lib/ext-plugin.lua:371]: I have added two test cases.

**Comment by rampagecong** [comment]
`[root@iZbp1efn5vn52xspp2bid5Z apisix]# prove -I. -I../test-nginx/inc -I../test-nginx/lib -r t/plugin/ext-plugin/http-req-call.t
t/plugin/ext-plugin/http-req-call.t .. ok    
All tests successful.
Files=1, Tests=18,  2 wallclock secs ( 0.02 usr  0.00 sys +  0.39 cusr  0.10 csys =  0.51 CPU)
Result: PASS`

Successfully!

File: apisix/plugins/ext-plugin/init.lua [hunk 7]
```diff
@@ -611,6 +611,29 @@ local rpc_handlers = {
                 end
             end
 
+            local len = rewrite:RespHeadersLength()
+            local exclude_resp_header = {
```
Comment (spacewander) [inline apisix/plugins/ext-plugin/init.lua:615]: Let's move the table in the module level

Comment (rampagecong) [reply apisix/plugins/ext-plugin/init.lua:615]: I understand to put it in the 'apisix/constants. lua' file, right?

Comment (shuaijinchao) [reply apisix/plugins/ext-plugin/init.lua:615]: You can move to the top of the current file.

File: apisix/plugins/ext-plugin/init.lua [hunk 8]
```diff
@@ -65,6 +65,18 @@ local type = type
...
+    ["connection"] = true,
+    ["content-length"] = true,
+    ["transfer-encoding"] = true,
+    ["location"] = true,
+    ["server"] = true,
+    ["www-authenticate"] = true,
+    ["content-encoding"] = true,
+    ["content-type"] = true,
+    ["content-location"] = true,
+    ["content-language"] = true,
```
Comment (tokers) [inline apisix/plugins/ext-plugin/init.lua:78]: Should we also exclude headers like: "Set-Cookie"?

Comment (tokers) [reply apisix/plugins/ext-plugin/init.lua:78]: Maybe some users want to use the external plugin to modify the Cookie, but it's also unsafe.

Comment (rampagecong) [reply apisix/plugins/ext-plugin/init.lua:78]: Yes,I want use the external plugin to modify the Cookie and there is a better way?

**Review by tzssangglass — APPROVED** [review]

**Review by spacewander — APPROVED** [review]

**Review by shuaijinchao — APPROVED** [review]

File: apisix/plugins/ext-plugin/init.lua (hunk 8 above)
Comment (spacewander) [reply apisix/plugins/ext-plugin/init.lua:78]: Maybe we can allow using  "Set-Cookie"? Setting Cookie won't break the upstream's response.

**Review by leslie-tsang — APPROVED** [review]

File: apisix/plugins/ext-plugin/init.lua (hunk 8 above)
Comment (tokers) [reply apisix/plugins/ext-plugin/init.lua:78]: I think we can just let it go and see the reaction of community.

**Pull request merged**